sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -58,6 +58,108 @@
 logger = logging.getLogger(__name__)
 
 
+class SnapshotCache(t.MutableMapping[SnapshotId, Snapshot]):
+    def __init__(self, cache: t.Optional[t.Dict[SnapshotId, Snapshot]] = None) -> None:
+        self._cache: t.Dict[SnapshotId, Snapshot] = {}
+        for k, v in (cache or {}).items():
+            self._set_snapshot(k, v)
+
+    def __getitem__(self, key: SnapshotId) -> Snapshot:
+        return self._cache[key]
+
+    def __setitem__(self, key: SnapshotId, value: Snapshot) -> None:
+        existing_snapshot = self.get(key)
+        if existing_snapshot:
+            value = value if value.updated_ts > existing_snapshot.updated_ts else existing_snapshot
+        self._set_snapshot(key, value)
+        self.merge_intervals_by_name_version()
+
+    def __delitem__(self, key: SnapshotId) -> None:
+        del self._cache[key]
+        self.merge_intervals_by_name_version()
+
+    def __iter__(self) -> t.Iterator[SnapshotId]:
+        return iter(self._cache)
+
+    def __len__(self) -> int:
+        return len(self._cache)
+
+    @property
+    def snapshots(self) -> t.List[Snapshot]:
+        return list(self.values())
+
+    @property
+    def is_hydrated(self) -> bool:
+        return all(snapshot.seed_is_hydrated for snapshot in self.values())
+
+    @property
+    def snapshot_intervals(self) -> t.List[SnapshotIntervals]:
+        return [snapshot.snapshot_intervals for snapshot in self.values()]
+
+    def get_cached_snapshots(self, snapshots: t.List[Snapshot]) -> t.List[Snapshot]:
```

Comment:
It seems counterintuitive that we mutate state in what appears to be a getter method. Should this be called `update_and_get...`?